### DISCUSSION TITLE
Passing Data to a script

### DISCUSSION THREAD

**Discussion by ifuller1** [shannonhochkins/ha-component-kit#74] [Q&A]
Hey all. Any idea how to pass data to a script entity? When I provide anything via "turnOn" it doesn't get executed at all. Runs fine without any data.

Cheers!

**Comment by shannonhochkins** [comment]
Have you tried using callService provided by useHass? If you have a custom script, the target must be the script name, ie script.script_name, then you can provide the service data serviceData object

**Comment by ifuller1** [comment]
No. That sounds perfect, thanks.

**Reply by shannonhochkins** [reply]
Let me know how you go!

**Reply by ifuller1** [reply]
Same as before. If I provide "serviceData" it doesn't work. I can see the call making it through to the socket connection so it suggests it's something about how HA accepts the request.

<img width="901" alt="Screenshot 2023-10-10 at 13 51 49" src="https://github.com/shannonhochkins/ha-component-kit/assets/144812/6677a7c6-216f-48f5-ada0-f071830e91b2">

**Comment by shannonhochkins** [comment]
Hi @ifuller1 , I just tried this and it worked fine but i can't see how your script is setup, an example of what i've just setup

```
alias: Say Something
sequence:
  - service: google_assistant_sdk.send_text_command
    data:
      media_player: media_player.office_hub
      command: "{{ message }}"
mode: single
icon: mdi:account-voice
```

This simple script, takes a dynamic message, then in `hakit`

```
const script = useEntity('script.say_something');
script.api.saySomething({
        message: 'What is 4 x 4'
});
```

And when i run the api.saySomething command it then replies on my hub with `4 x 4 is 16` so it's all working fine, keep in mind that im not using `turnOn` here, if you're using the typescript sync functionality i've built, ensure you're updated with your home assistant, because after i created the script in HA, i ran the sync and the intellisense gave me `saySomething` on the api object which is what worked for me (the script name)

Additionally, if you're trying to debug web sockets, there's a much easier way using the developer tools, refresh the page, you'll see a websocket/api connection, if you click on that, then messages, you'll see all socket requests funneling through:

![Screenshot 2023-10-11 at 8 21 40 am](https://github.com/shannonhochkins/ha-component-kit/assets/5105002/e8cd9c61-b0ad-4f99-b3ca-8f3491bd4307)

**Comment by ifuller1** [comment]
I'll try this exact setup thanks.

I was aware of the web socket debug although I probably should have gone
straight there!

Thanks

**Reply by shannonhochkins** [reply]
If you get stuck, might be helpful to see your script yaml from home assistant, however this did work for me fine :)

**Comment by ifuller1** [comment]
So I copied your exact script. Strangely it doesn't trigger at all (even without data being provided). I can see the service call being made (and have double checked the entity ID). 

I did go back to my previous script (just in case it was something specific to the script) and that runs fine now with data using the `script.api.camelCaseName`!

Couple of screenshots here  if they show anything obvious. The "say_something" script appears to get an "off", "off" response.

I guess my use case is working but it would be good to understand why/how! :)

<img width="1045" alt="Screenshot 2023-10-11 at 10 05 39" src="https://github.com/shannonhochkins/ha-component-kit/assets/144812/754516b8-aaf1-4ab0-9637-049b7491ca56">

<img width="1045" alt="Screenshot 2023-10-11 at 09 59 11" src="https://github.com/shannonhochkins/ha-component-kit/assets/144812/ffc4baea-2679-4c23-95c7-9adf6eca9ffb">


<img width="582" alt="Screenshot 2023-10-11 at 09 58 26" src="https://github.com/shannonhochkins/ha-component-kit/assets/144812/52c8174e-ca70-4aea-b9c4-7e12777335ba">

<img width="635" alt="Screenshot 2023-10-11 at 10 09 12" src="https://github.com/shannonhochkins/ha-component-kit/assets/144812/d76ad4c7-cf4e-4d4e-b720-1fa831f5a2a9">

p.s. what is the `api.camelCaseName` format? Can I just take the snake case id and camel case it?

**Reply by shannonhochkins** [reply]
I'm a little confused about your questions here  you said it didn't work, but then it did work?

I am not sure if you have the same google sdk addon installed, if you don't that simply won't work

Regarding the names of the method, you're correct snake case has been converted to camel case for the service names and domain names, if you have typescript installed with the sync script you'll get full intellisense of what's available including the arguments for services 😉

**Comment by ifuller1** [comment]
So calling your script didn't work at all - it didn't get called with or
without the params. (see the screenshot - not sure if anything obvious in
the web socket messages).

Calling my script, using this syntax, is working - so I guess I'm good!

**Reply by shannonhochkins** [reply]
That's odd, but I'm glad you got it working!

**Comment by ifuller1** [comment]
I recently updated HA so not sure if it's related. I assume the response
messages mean nothing to you?

**Reply by shannonhochkins** [reply]
The response of on or off? It's simply just indicating that it was triggered I believe, but probably best to check ha documentation, is it causing you problems? You can simply ignore the response if it's all working

**Reply by shannonhochkins** [reply]
The response of on or off? It's simply just indicating that it was triggered I believe, but probably best to check ha documentation, is it causing you problems? You can simply ignore the response if it's all working

**Comment by ifuller1** [comment]
It's just so strange as you can see it receives the messages but ha says 0
runs.

Anywho. Thanks!

**Reply by shannonhochkins** [reply]
It's entirely possible that the runs information displayed is some higher order function to save in the database, if you run the script from HA and inspect the websockets, you could confirm that the same request is being made as hakit and that'll confirm that its doing something else based on the response to update that runs value

**Comment by ifuller1** [comment]
Good shout. Thanks
